numpy/lib/function_base.py

```diff
@@ -1904,10 +1927,89 @@ def diff(a, n=1, axis=-1):
     slice1 = tuple(slice1)
     slice2 = tuple(slice2)
     if n > 1:
-        return diff(a[slice1]-a[slice2], n-1, axis=axis)
-    else:
+        return diff(a[slice1]-a[slice2], n-1, axis=axis, 
+                    prepend=prepend, append=append)
+
+    # if nothing to add to either side, fast track the result
+    elif prepend is np._NoValue and append is np._NoValue:
         return a[slice1]-a[slice2]
 
+    else:
+        # logic to handle append input
+        # l_end is the length of padding at the end
+        if append is np._NoValue:
+            l_end = 0
+        elif isinstance(append, str) and append == 'last':
+            if a.shape[axis] != 0:
+                append = a.take([-1], axis)
+                l_end = 1
+            else:
+                l_end = 0
+        else:
+            append = np.asanyarray(append)
+            if append.ndim == 0:
+                l_end = 1
+            elif append.ndim != a.ndim:
+                raise ValueError('append must have the same dimension as '
+                                 'the input array')
+            else:
+                l_end = append.shape[axis]
+
+        # logic to handle prepend input
+        # l_begin is the length of padding at the beginning
+        if prepend is np._NoValue:
+            l_begin = 0
+        elif isinstance(prepend, str) and prepend == 'first':
```

Comment:
I think it's probably safer to outright forbid any other string values, rather than risking some weird conversion because someone mistyped `first`. If users really want to append a string scalar, they can use `np.array('first')`, which also has the bonus of allowing them to append a literal `'first'`